internal/engine/pod_watch.go

```diff
@@ -60,36 +75,57 @@ func subtract(a, b []PodWatch) []PodWatch {
 	return ret
 }
 
-func (w *PodWatcher) diff(ctx context.Context, st store.RStore) (setup []PodWatch, teardown []PodWatch) {
+type podWatchTaskList struct {
+	setup    []PodWatch
+	teardown []PodWatch
+	newUIDs  map[types.UID]model.ManifestName
+}
+
+func (w *PodWatcher) diff(ctx context.Context, st store.RStore) podWatchTaskList {
 	state := st.RLockState()
 	defer st.RUnlockState()
 
 	w.mu.RLock()
 	defer w.mu.RUnlock()
 
 	atLeastOneK8s := false
+	newUIDs := make(map[types.UID]model.ManifestName)
 	var neededWatches []PodWatch
-	for _, m := range state.Manifests() {
-		if m.IsK8s() {
-			atLeastOneK8s = true
-			for _, ls := range m.K8sTarget().ExtraPodSelectors {
-				if !ls.Empty() {
-					neededWatches = append(neededWatches, PodWatch{name: m.Name, labels: ls})
-				}
+	for _, mt := range state.Targets() {
+		if !mt.Manifest.IsK8s() {
+			continue
+		}
+
+		name := mt.Manifest.Name
+		atLeastOneK8s = true
+		for _, ls := range mt.Manifest.K8sTarget().ExtraPodSelectors {
+			if !ls.Empty() {
+				neededWatches = append(neededWatches, PodWatch{name: mt.Manifest.Name, labels: ls})
+			}
+		}
+
+		for id := range mt.State.K8sRuntimeState().DeployedUIDSet {
+			oldName := w.knownDeployedUIDs[id]
```

Comment:
more comments pls, for the sake of the next person to work in this file? esp. here, but throughout

Reply:
done